### PR TITLE
Added test-case for deprecated phpstorm stub with patch-version

### PR DESCRIPTION
**Pull request by staabm**
tests the ternary in https://github.com/Roave/BetterReflection/blob/ab47d0b335a33ac211d5ad8758a48ed6be8622ca/src/SourceLocator/SourceStubber/PhpStormStubsSourceStubber.php#L677 which wasn't covered before

### CODE REVIEW

**Comment by Ocramius** [comment]
> which wasn't covered before

Was it an escaped / uncovered mutation?

**Comment by staabm** [comment]
> Was it an escaped / uncovered mutation?

I don't know. I am not a mutation testing expert. 

but I realized that this is a path which wasn't test-covered before while working on the PHPStan integration

**Comment by Ocramius** [comment]
> I don't know. I am not a mutation testing expert.

I was more wondering if mainline caught it - should be visible in the MT framework output :D

**Comment by Ocramius** [comment]
(wrong button - thanks Github for not dong TAB+ENTER correctly)

**Comment by kukulich** [comment]
https://github.com/Roave/BetterReflection/commit/28ef29b9e81c77a3157a33af64486e5c1545fa3c

So I think there may be some escaped mutants now...

**Comment by Ocramius** [comment]
FWIW, in [mainline](https://github.com/Roave/BetterReflection/actions/runs/8259610533/job/22593799759):

```

6) /home/runner/work/BetterReflection/BetterReflection/src/SourceLocator/SourceStubber/PhpStormStubsSourceStubber.php:676    [M] IncrementInteger

--- Original
+++ New
@@ @@
         }
         if (preg_match('#@deprecated\\s+(\\d+)\\.(\\d+)(?:\\.(\\d+)?)$#m', $docComment->getText(), $matches) === 1) {
             $major = $matches[1];
-            $minor = $matches[2];
+            $minor = $matches[3];
             $patch = $matches[3] ?? 0;
             $versionId = sprintf('%d%02d%02d', $major, $minor, $patch);
             if ($this->phpVersion >= $versionId) {


7) /home/runner/work/BetterReflection/BetterReflection/src/SourceLocator/SourceStubber/PhpStormStubsSourceStubber.php:677    [M] DecrementInteger

--- Original
+++ New
@@ @@
         if (preg_match('#@deprecated\\s+(\\d+)\\.(\\d+)(?:\\.(\\d+)?)$#m', $docComment->getText(), $matches) === 1) {
             $major = $matches[1];
             $minor = $matches[2];
-            $patch = $matches[3] ?? 0;
+            $patch = $matches[2] ?? 0;
             $versionId = sprintf('%d%02d%02d', $major, $minor, $patch);
             if ($this->phpVersion >= $versionId) {
                 return true;


8) /home/runner/work/BetterReflection/BetterReflection/src/SourceLocator/SourceStubber/PhpStormStubsSourceStubber.php:677    [M] IncrementInteger

--- Original
+++ New
@@ @@
         if (preg_match('#@deprecated\\s+(\\d+)\\.(\\d+)(?:\\.(\\d+)?)$#m', $docComment->getText(), $matches) === 1) {
             $major = $matches[1];
             $minor = $matches[2];
-            $patch = $matches[3] ?? 0;
+            $patch = $matches[4] ?? 0;
             $versionId = sprintf('%d%02d%02d', $major, $minor, $patch);
             if ($this->phpVersion >= $versionId) {
                 return true;


9) /home/runner/work/BetterReflection/BetterReflection/src/SourceLocator/SourceStubber/PhpStormStubsSourceStubber.php:677    [M] DecrementInteger

--- Original
+++ New
@@ @@
         if (preg_match('#@deprecated\\s+(\\d+)\\.(\\d+)(?:\\.(\\d+)?)$#m', $docComment->getText(), $matches) === 1) {
             $major = $matches[1];
             $minor = $matches[2];
-            $patch = $matches[3] ?? 0;
+            $patch = $matches[3] ?? -1;
             $versionId = sprintf('%d%02d%02d', $major, $minor, $patch);
             if ($this->phpVersion >= $versionId) {
                 return true;


10) /home/runner/work/BetterReflection/BetterReflection/src/SourceLocator/SourceStubber/PhpStormStubsSourceStubber.php:677    [M] IncrementInteger

--- Original
+++ New
@@ @@
         if (preg_match('#@deprecated\\s+(\\d+)\\.(\\d+)(?:\\.(\\d+)?)$#m', $docComment->getText(), $matches) === 1) {
             $major = $matches[1];
             $minor = $matches[2];
-            $patch = $matches[3] ?? 0;
+            $patch = $matches[3] ?? 1;
             $versionId = sprintf('%d%02d%02d', $major, $minor, $patch);
             if ($this->phpVersion >= $versionId) {
                 return true;


11) /home/runner/work/BetterReflection/BetterReflection/src/SourceLocator/SourceStubber/PhpStormStubsSourceStubber.php:677    [M] Coalesce

--- Original
+++ New
@@ @@
         if (preg_match('#@deprecated\\s+(\\d+)\\.(\\d+)(?:\\.(\\d+)?)$#m', $docComment->getText(), $matches) === 1) {
             $major = $matches[1];
             $minor = $matches[2];
-            $patch = $matches[3] ?? 0;
+            $patch = 0 ?? $matches[3];
             $versionId = sprintf('%d%02d%02d', $major, $minor, $patch);
             if ($this->phpVersion >= $versionId) {
                 return true;
```

That should be OK with this patch now :) 

This is mostly to show @staabm how to read this: I think MT is something that will be great for your testing toolbox, since you are so detail-oriented :muscle:

**Comment by Ocramius** [comment]
I don't see mutation changes here: I think your test scenario did **not** hit the edge case you were going for, as the mutations stayed the same :D

**Comment by staabm** [comment]
thanks for the explanation.

lets see whether we can kill some mainline mutants with https://github.com/nikic/PHP-Parser/pull/985 :-)

**Comment by staabm** [comment]
> FWIW, in [mainline](https://github.com/Roave/BetterReflection/actions/runs/8259610533/job/22593799759):

looking at the latest build of the latest `6.28.x`-commit, I don't see the mutants you referenced in your comment

https://github.com/Roave/BetterReflection/actions/runs/8259610533/job/22593799759

File: test/unit/SourceLocator/SourceStubber/PhpStormStubsSourceStubberTest.php [hunk 1]
```diff
@@ -501,6 +501,83 @@ public function testStubForConstantThatIsDeprecated(): void
...
+\define('A_CUSTOM_CONSTANT', 513);
+EOT;
+        $stubData    = new StubData($exampleStub, null);
+
+        self::assertStringMatchesFormat(
+            "%Adefine('A_CUSTOM_CONSTANT',%w%d);",
+            $stubData->getStub(),
+        );
+
+        if (PHP_VERSION_ID >= 80100) {
```
Comment (staabm) [inline test/unit/SourceLocator/SourceStubber/PhpStormStubsSourceStubberTest.php:522]: ahh I think we don't catch a bug because we never reach the ELSE case of this condition (because CI is running only 8.2/8.3)

Comment (Ocramius) [reply test/unit/SourceLocator/SourceStubber/PhpStormStubsSourceStubberTest.php:522]: Ah, good catch! Indeed! 

Perhaps we can change the fixture to introduce deprecations with higher/fictional PHP versions? :D

Comment (staabm) [reply test/unit/SourceLocator/SourceStubber/PhpStormStubsSourceStubberTest.php:522]: Let me first make sure phpstan can catch such bugs in the future. Brb

Comment (staabm) [reply test/unit/SourceLocator/SourceStubber/PhpStormStubsSourceStubberTest.php:522]: https://github.com/phpstan/phpstan-src/pull/2968 should help us in the future.

**Comment by staabm** [comment]
superseded by https://github.com/Roave/BetterReflection/pull/1408
